tests/main/snap-session-agent-unavailable-to-snaps/task.yaml

```diff
@@ -0,0 +1,55 @@
+summary: Check that snaps can not access the session agent
+
+systems:
+    # Ubuntu 14.04 does not have a complete systemd implementation
+    - -ubuntu-14.04-*
+    # Systemd on CentOS 7/Amazon Linux 2 does not have the user@uid unit
+    - -amazon-linux-2-*
+    - -centos-7-*
+
+environment:
+    TEST_UID: $(id -u test)
+    USER_RUNTIME_DIR: /run/user/${TEST_UID}
+
+prepare: |
+    # Ensure that snapd.session-agent.socket is enabled.  This may not
+    # be the case on distributions where presets have been used to
+    # disable it.
+    if [ ! -L /usr/lib/systemd/user/sockets.target.wants/snapd.session-agent.socket ] &&
+            ! systemctl --user --global is-enabled snapd.session-agent.socket; then
+        systemctl --user --global enable snapd.session-agent.socket
+        touch agent-was-enabled
+    fi
+    mkdir -p "$USER_RUNTIME_DIR"
+    chmod u=rwX,go= "$USER_RUNTIME_DIR"
+    chown test:test "$USER_RUNTIME_DIR"
+    systemctl start "user@${TEST_UID}.service"
+
+restore: |
+    systemctl stop "user@${TEST_UID}.service"
+    rm -rf "${USER_RUNTIME_DIR:?}"/* "${USER_RUNTIME_DIR:?}"/.[!.]*
+    if [ -f agent-was-enabled ]; then
+        systemctl --user --global disable snapd.session-agent.socket
+        rm agent-was-enabled
+    fi
+
+execute: |
+    if [ "$(snap debug confinement)" != strict ]; then
```

Comment:
Instead of just looking at strict, we should really probe for the exact sandbox feature. OTOH, comparing sandbox-features of apparmor backend on Ubuntu and Arch, we need to make some updates, so this bit of code is fine as it is.

Reply:
I take that back. I think we could use a test like this here: 
```
if ! snap debug sandbox-features --required apparmor:policy:default --required apparmor:kernel:file; then
    exit 0
fi
```
And the test should be able to complete on Ubuntu and on Arch.

Reply:
Played with the snippet I proposed and it fails on Arch. Looks like unix socket mediation, even path based, is not covered by `file` feature. Therefore, the user access to the socket is not blocked.

On the upside, root access to user's socket is blocked, but only as a side effect of  `dac_override` being denied.

Reply:
> Played with the snippet I proposed and it fails on Arch. Looks like unix socket mediation, even path based, is not covered by `file` feature. Therefore, the user access to the socket is not blocked.

Named sockets have been handled since... forever in apparmor. Only anonymous and abstract sockets require the unix mediation patches. Did the process go into forced devmode (look in /var/lib/snapd/apparmor/profiles/snap.your.thing)? I was testing this on debian sid with hello-world and it is falling back to forced devmode:
```
Dec  5 07:37:13 debian-sid-amd64 snapd[1054]: AppArmor status: apparmor is enabled but some kernel features are missing: dbus, network
Dec  5 07:37:13 debian-sid-amd64 snapd[1054]: daemon.go:346: started snapd/2.42.1-1 (series 16; classic; devmode) debian/ (amd64) linux/5.3.0-2-amd64.
```

However, if I copy the profile from an Ubuntu system I can see that file mediation is sufficient to mediate access to the socket:
```
$ scp /var/lib/snapd/apparmor/profiles/snap.hello-world.sh debian-sid-amd64:/tmp
snap.hello-world.sh                                100%   18KB  12.2MB/s   00:00
$ ssh debian-sid-amd64

jamie@debian-sid-amd64:~$ sudo apparmor_parser -r /tmp/snap.hello-world.sh

jamie@debian-sid-amd64:~$ ls /sys/kernel/security/apparmor/features/
capability  domain  mount       network_v8  ptrace  rlimit
caps        file    namespaces  policy      query   signal

jamie@debian-sid-amd64:~$ sudo aa-exec -p snap.hello-world.sh -- head /run/snapd.socket
head: cannot open '/run/snapd.socket' for reading: Permission denied

jamie@debian-sid-amd64:~$ sudo grep DEN /var/log/syslog|tail -1
Dec  5 07:49:26 debian-sid-amd64 kernel: [  837.318954] audit: type=1400 audit(1575553766.233:21): apparmor="DENIED" operation="open" profile="snap.hello-world.sh" name="/run/snapd.socket" pid=1451 comm="head" requested_mask="r" denied_mask="r" fsuid=0 ouid=0
```

Reply:
> Named sockets have been handled since... forever in apparmor.

@bboozzoo - ok, I discussed this with jj on IRC and named socket mediation is complicated and your black box assessment was correct. 

First off, https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/include/linux/lsm_hooks.h#n761 is out of date.

Second, while historically, connect() was mediated by the file hooks via inode hooks (ie, what I remembered), at some point in the now fairly distant past, the hooks to mediate connect() were moved into unix_stream_connect (and unix_may_send) hooks in af_unix (eg, 12.04 LTS has these changes). At this point, AppArmor lost mediation of connect() for named sockets entirely.

Third, fast forward to when network_v8 was upstreamed, AppArmor was able to mediate named sockets again via rules like 'network af_unix stream', but network_v8 is not usable without apparmor 3 userspace (which isn't released yet, but planned this cycle).

The still out-of-tree but getting closer af_unix patchset (eg, [this one for 4.17](https://gitlab.com/apparmor/apparmor/blob/master/kernel-patches/v4.17/0002-apparmor-af_unix-mediation.patch#L964)) implements the unix_stream_connect() and unix_may_send() hooks in such a manner that 'file' rules mediate connect() again for named sockets.

Sorry for the noise, but hopefully it is useful understanding the historical context.

Reply:
@jdstrand thank you, this is super insightful!